spring-xd-dirt/src/main/java/org/springframework/xd/dirt/module/ResourceModuleRegistry.java

```diff
@@ -16,25 +16,319 @@
+					|| resource instanceof FileSystemResource) {
+				resource = new FileSystemResource(resource.getFile().getCanonicalFile());
+			}
+			filtered.add(resource);
+		}
+		return filtered;
+	}
+
+	protected void collect(Resource resource, List<ModuleDefinition> holder) throws IOException {
+		// ResourcePatternResolver.getResources() may return resources that don't exist when not using wildcards
+		if (!resource.exists()) {
+			return;
+		}
+
+		String path = resource.getURI().toString();
+		if (path.endsWith(HASH_EXTENSION)) {
+			return;
+		}
+		else if (path.endsWith(ARCHIVE_AS_FILE_EXTENSION) && requireHashFiles && !hashResource(resource).exists()) {
+			return;
+		}
+		// URI paths for directories include an extra slash, strip it for now
+		if (path.endsWith("/")) {
+			path = path.substring(0, path.length() - 1);
+		}
+
+		int lastSlash = path.lastIndexOf('/');
+		int nextToLastSlash = path.lastIndexOf('/', lastSlash - 1);
+
+		String name = path.substring(lastSlash + 1);
+		if (name.endsWith(ARCHIVE_AS_FILE_EXTENSION)) {
+			name = name.substring(0, name.length() - ARCHIVE_AS_FILE_EXTENSION.length());
+		}
+		String typeAsString = path.substring(nextToLastSlash + 1, lastSlash);
+		ModuleType type = null;
+		try {
+			type = ModuleType.valueOf(typeAsString);
+		}
+		catch (IllegalArgumentException e) {
+			// Not an actual type name, skip
+			return;
+		}
+		String locationToUse = resource.getURL().toString();
+
+		ModuleDefinition found = ModuleDefinitions.simple(name, type, locationToUse);
+
+		if (holder.contains(found)) {
+			SimpleModuleDefinition one = (SimpleModuleDefinition) found;
+			SimpleModuleDefinition two = (SimpleModuleDefinition) holder.get(holder.indexOf(found));
+			throw new IllegalStateException(String.format("Duplicate module definitions for '%s:%s' found at '%s' and '%s'",
+					found.getType(), found.getName(), one.getLocation(), two.getLocation()));
+		}
+		else {
+			holder.add(found);
+		}
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		if (root.startsWith("hdfs:")) {
+			ConfigurationFactoryBean configurationFactoryBean = new ConfigurationFactoryBean();
+			configurationFactoryBean.setRegisterUrlHandler(true);
+			configurationFactoryBean.setFileSystemUri(root);
+			configurationFactoryBean.afterPropertiesSet();
+
+			this.resolver = new HdfsResourceLoader(configurationFactoryBean.getObject());
+		}
+
+		if (createDirectoryStructure) {
+			// Create intermediary folders
+			for (ModuleType type : ModuleType.values()) {
+				Resource folder = getResources(type.name(), "", "").iterator().next();
+				if (!folder.exists()) {
+					ExtendedResource.wrap(folder).mkdirs();
+				}
+			}
+		}
+
+	}
+
+	public void setRequireHashFiles(boolean requireHashFiles) {
+		this.requireHashFiles = requireHashFiles;
+	}
+
+	public void setCreateDirectoryStructure(boolean createDirectoryStructure) {
+		this.createDirectoryStructure = createDirectoryStructure;
+	}
```

Comment:
Do we really need to make these properties independently configurable? Seems like it could lead to some inconsistencies. I don't see these setters used anywhere.